File: client/lib/src/envelope/envelope_rejection.dart
/// Base class for all envelope verification failures.
sealed class EnvelopeRejection implements Exception {
  final String message;

  const EnvelopeRejection(this.message);

  @override
  String toString() => '$runtimeType: $message';
}

/// The CBOR bytes could not be parsed or are not in canonical form.
final class MalformedCbor extends EnvelopeRejection {
  final Object? cause;

  const MalformedCbor(String message, [this.cause]) : super(message);
}

/// The envelope version is not supported by this implementation.
final class UnsupportedVersion extends EnvelopeRejection {
  UnsupportedVersion(int version)
      : super('Unsupported envelope version: $version');
}

/// The issuer signing key was not active at [createdAt].
final class IssuerKeyNotActiveAt extends EnvelopeRejection {
  const IssuerKeyNotActiveAt(String detail) : super(detail);
}

/// The Ed25519 signature did not verify.
final class SignatureInvalid extends EnvelopeRejection {
  const SignatureInvalid() : super('Envelope signature verification failed');
}

/// The SHA-256 of the ciphertext does not match [ciphertext_hash].
final class CiphertextHashMismatch extends EnvelopeRejection {
  const CiphertextHashMismatch()
      : super('SHA-256 of ciphertext does not match ciphertext_hash');
}

File: src/main/java/com/wilhelmsen/cbslink/plugin/datawallet/envelope/EnvelopeRejection.java
package com.wilhelmsen.cbslink.plugin.datawallet.envelope;

public sealed class EnvelopeRejection extends RuntimeException {

    EnvelopeRejection(String message) {
        super(message);
    }

    EnvelopeRejection(String message, Throwable cause) {
        super(message, cause);
    }

    public static final class MalformedCbor extends EnvelopeRejection {
        public MalformedCbor(String message) { super(message); }
        public MalformedCbor(String message, Throwable cause) { super(message, cause); }
    }

    public static final class UnsupportedVersion extends EnvelopeRejection {
        public UnsupportedVersion(int version) {
            super("Unsupported envelope version: " + version);
        }
    }

    public static final class SignatureInvalid extends EnvelopeRejection {
        public SignatureInvalid() { super("Envelope signature verification failed"); }
    }

    public static final class IssuerKeyNotActiveAt extends EnvelopeRejection {
        public IssuerKeyNotActiveAt(String detail) { super(detail); }
    }

    public static final class CiphertextHashMismatch extends EnvelopeRejection {
        public CiphertextHashMismatch() { super("SHA-256 of ciphertext does not match ciphertext_hash"); }
    }

    public static final class RecipientNotPresent extends EnvelopeRejection {
        public RecipientNotPresent(String detail) { super(detail); }
    }
}
